libusb/os/windows_winusb.c

```diff
@@ -2017,6 +2017,157 @@ static int winusb_get_device_list(struct libusb_context *ctx, struct discovered_
 	return r;
 }
 
+struct string_descriptor_s {
+	UCHAR bLength;
+	UCHAR bDescriptorType;
+	uint8_t bString[252];  // UTF-16LE, must be even length
+};
+
+struct string_descriptor_req_s {
+	USB_DESCRIPTOR_REQUEST req;
+	struct string_descriptor_s desc;
+};
+
+static int usbi_utf16le_to_utf8(uint8_t const* src, int src_length, char* dst, int dst_length) {
```

Comment:
I'm not a fan having hand-written UTF encoding handling (unless it is a iconv sources).
Have you considered using `WideCharToMultiByte` instead?

Reply:
I really did not want to introduce any new dependencies, so not iconv.

I am not a fan of the Microsoft UTF8 handling.  The [docs](https://learn.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-widechartomultibyte) say this:
_Caution  Using the WideCharToMultiByte function incorrectly can compromise the security of your application. Calling this function can easily cause a buffer overrun because the size of the input buffer indicated by lpWideCharStr equals the number of characters in the Unicode string, while the size of the output buffer indicated by lpMultiByteStr equals the number of bytes. To avoid a buffer overrun, your application must specify a buffer size appropriate for the data type the buffer receives._

In contrast, this function properly truncates on overflow.  With the way we currently call this with length 384 strings, this is not an issue, but it could be if anyone "properly" uses the usbi API in the future.

This function comes from another project of mine where I am the copyright holder, so I can contribute it under LGPL. 
 Here is the unit test:

```
TEST_DEF(utf16le_to_utf8) {
    uint8_t src[] = {'a', 0, 'b', 0, 0x80, 0x00, 0x00, 0x08, 0x00, 0x00};
    uint8_t dst_expect[] = {'a', 'b', 0xC2, 0x80, 0xE0, 0xA0, 0x80, 0x00};

    char dst[8];
    assert_eq_u32(8, os_cstr_utf16le_to_utf8(src, NULL, 0));

    assert_eq_u32(8, os_cstr_utf16le_to_utf8(src, dst, 8));
    assert_eq_mem(dst_expect, 8, dst, 8);

    uint8_t dst_expect2[] = {'a', 'b', 0};
    assert_eq_u32(8, os_cstr_utf16le_to_utf8(src, dst, 4));
    assert_eq_mem(dst_expect2, 3, dst, 3);
}
```

This really belongs in it's own file along with the utf8 string copy.  However, now I would have to add files to Make, XCode and msvc builds correctly  😬

Reply:
In this case, at least add a comment/reference where this came from (if publicly available).

I suspect @tormodvolden or @hjelmn might have oppinion to where is best to place this function if not keep it right here.

Reply:
Unfortunately, where I yanked this from is not publicly available.  The source for this used for the same purpose.  I went back and modified it to handle both null-terminated and length-terminated UTF-16LE strings.  I will push the revised code soon.  Here is the passing unit test:
```
TEST_DEF(utf16le_to_utf8) {
    uint8_t src[] = {'a', 0, 'b', 0, 0x80, 0x00, 0x00, 0x08, 0x00, 0x00};
    uint8_t dst_expect[] = {'a', 'b', 0xC2, 0x80, 0xE0, 0xA0, 0x80, 0x00};

    char dst[8];
    dst[0] = '!';
    assert_eq_u32(0, os_cstr_utf16le_to_utf8(NULL, 0, dst, sizeof(dst)));
    assert_eq_u32(0, dst[0]);

    assert_eq_u32(8, os_cstr_utf16le_to_utf8(src, sizeof(src), NULL, 0));
    assert_eq_u32(8, os_cstr_utf16le_to_utf8(src, sizeof(src) - 2, NULL, 0));  // no null termination

    assert_eq_u32(8, os_cstr_utf16le_to_utf8(src, sizeof(src), dst, 8));
    assert_eq_mem(dst_expect, 8, dst, 8);

    uint8_t dst_expect2[] = {'a', 'b', 0};
    assert_eq_u32(8, os_cstr_utf16le_to_utf8(src, sizeof(src), dst, 4));
    assert_eq_mem(dst_expect2, 3, dst, 3);
}
```

If this were my codebase to maintain, I would want to put usbi_utf16le_to_utf8() and the upcoming usbi_utf8_copy() into their own file with their own unit tests.  I am not comfortable modifying the existing libusb build system across all the platforms to make this happen.  CMake helps solve this problem...

Reply:
> I am not comfortable modifying the existing libusb build system across all the platforms to make this happen

That's alright. But at least add a comment describing where it is from, even if you cannot add a link/reference.